framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -4631,6 +4613,442 @@ private void makeGroundTargetType(
+              (AnnotatedTypeVariable) capturedTypeArg);
+          newTypeArgs.add(capturedTypeArg);
+        } else {
+          // If the bounds of the captured wildcard are the same, then it is converted to
+          // a declared type. This seems to be a violation of the JLS, but javac does
+          // this, so the Checker Framework must do it.
+          typeVariableMapping.put(typeVarTypeMirror, capturedTypeArg);
+          newTypeArgs.add(capturedTypeArg);
+          replaceAnnotations(
+              ((AnnotatedWildcardType) uncapturedTypeArg).getSuperBound(), capturedTypeArg);
+        }
+      } else {
+        // The type argument is not a captured type.
+        typeVariableMapping.put(typeVarTypeMirror, uncapturedTypeArg);
+        if (uncapturedTypeArg.getKind() == TypeKind.TYPEVAR) {
+          AnnotatedTypeVariable typeVar = (AnnotatedTypeVariable) uncapturedTypeArg;
+          typeVariableMapping.put(typeVar.getUnderlyingType(), typeVar);
+        }
+        newTypeArgs.add(uncapturedTypeArg);
+      }
+    }
+
+    // Loop through the type arguments and set the annotations of each captured type.
+    List<AnnotatedTypeVariable> orderToCapture = order(capturedTypeMapping.values());
+    for (AnnotatedTypeVariable capturedTypeArg : orderToCapture) {
+      int i = capturedTypeMirror.getTypeArguments().indexOf(capturedTypeArg.getUnderlyingType());
+      AnnotatedTypeMirror uncapturedTypeArg = uncapturedType.getTypeArguments().get(i);
+      AnnotatedTypeVariable typeVariable =
+          (AnnotatedTypeVariable) typeDeclaration.getTypeArguments().get(i);
+      captureWildcard(
+          typeVariableMapping,
+          capturedTypeMapping,
+          (AnnotatedWildcardType) uncapturedTypeArg,
+          typeVariable,
+          capturedTypeArg);
+      newTypeArgs.remove(i);
+      newTypeArgs.add(i, capturedTypeArg);
+    }
+
+    capturedType.setTypeArguments(newTypeArgs);
+    // Use the mapping above to substitute the type variables in capturedType.
+
+    capturedType.addAnnotations(uncapturedType.getAnnotations());
+    return capturedType;
+  }
+
+  /**
+   * Copy the non-wildcard type args from a uncapturedType to its captureType. Also, ensure that
+   * type variables in capturedType are the same object when they are refer to the same type
+   * variable.
+   *
+   * <p>To use, Call {@code #copy(AnnotatedDeclaredType, AnnotatedDeclaredType)} rather than a visit
+   * method.
+   */
+  private final CaptureTypeArgCopier captureTypeArgCopier = new CaptureTypeArgCopier();
+
+  /**
+   * Copy the non-wildcard type args from {@code uncapturedType} to {@code captureType}. Also,
+   * ensure that type variables in {@code capturedType} are the same object when they are refer to
+   * the same type variable.
+   *
+   * <p>To use, Call {@code #copy(AnnotatedDeclaredType, AnnotatedDeclaredType)} rather than a visit
+   * method.
+   */
+  private class CaptureTypeArgCopier extends AnnotatedTypeCopier {
```

Comment:
Maybe it would be more descriptive to name this `copyNonWildcardTypeArgs` or `NonWildcardTypeArgCopier`, which describes what it does rather than where it is used.